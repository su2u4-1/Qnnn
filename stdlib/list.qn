class list<T> {
    attr int size, real_size;
    attr arr<T> c;

    method op void init(self, int size, T value) {
        self.size = size;
        self.real_size = size + 4;
        self.c = arr<T>(size + 4);
        for (int i in range(size)) {
            self.c[i] = value;
        }
        return;
    }

    method op void init(self, arr<T> a) {
        self.c = a;
        self.size = a.size();
        return;
    }

    method op T get(self, int index) {
        if (index < 0 || index >= self.size) {
            raise(Except("Index out of bounds"));
        }
        return self.c[index];
    }

    method op void set(self, int index, T value) {
        if (index < 0 || index >= self.size) {
            raise(Except("Index out of bounds"));
        }
        self.c[index] = value;
        return;
    }

    method op bool eq(self, list<T> other) {
        if (self.size != other.size()) {
            return false;
        }
        for (int i in range(self.size)) {
            if (self.c[i] != other.c[i]) {
                return false;
            }
        }
        return true;
    }

    method op list<T> mul(self, int n) {
        var arr<T> t = arr<T>(self.size * n);
        for (int i in range(n)) {
            for (int j in range(self.size)) {
                t[i * n + j] = self.c[j];
            }
        }
        return list<T>(t);
    }

    method op void imul(self, int n) {
        var arr<T> t = arr<T>(self.size * n + 4);
        for (int i in range(n)) {
            for (int j in range(self.size)) {
                t[i * n + j] = self.c[j];
            }
        }
        self.c = t;
        self.size *= n;
        self.real_size = self.size + 4;
        return self;
    }

    method op list<T> add(self, list<T> other) {
        var arr<T> t = arr<T>(self.size + other.size());
        for (int i in range(self.size)) {
            t[i] = self.c[i];
        }
        for (int i in range(other.size)) {
            t[self.size + i] = other.c[i];
        }
        return list<T>(t);
    }

    method op void iadd(self, list<T> other) {
        self.extend(other);
        return;
    }

    method op string str(self) {
        var string result = "[";
        for (T i in self.c) {
            if (i == self.c[self.size - 1]) {
                result += string(i) + "]";
            } else {
                result += string(i) + ", ";
            }
        }
        return result;
    }

    method public void append(self, T value) {
        self.size += 1;
        if (self.size >= self.real_size) {
            var arr<T> t = arr<T>(self.real_size * 2);
            for (int i in range(self.size)) {
                t[i] = self.c[i];
            }
            self.c = t;
            self.real_size *= 2;
        }
        self.c[self.size - 1] = value;
        return;
    }

    method public int size(self) {
        return self.size;
    }

    method public void clear(self) {
        self.c = arr<T>(4);
        self.size = 0;
        self.real_size = 4;
        return;
    }

    method public void insert(self, int n, T value) {
        if (n < 0 || n > self.size) {
            raise(Except("Index out of bounds"));
        }
        self.size += 1;
        if (self.size >= self.real_size) {
            var arr<T> t = arr<T>(self.real_size * 2);
            for (int i in range(n)) {
                t[i] = self.c[i];
            }
            self.c = t;
            self.real_size *= 2;
        }
        var T t1, t2 = value;
        for (int i in range(n, self.size)) {
            t1 = self.c[i];
            self.c[i] = t2;
            t2 = t1;
        }
        return;
    }

    method public void extend(self, list<T> other) {
        if (self.real_size < self.size + other.size) {
            var arr<T> t = arr<T>(self.size + other.size() + 4);
            for (int i in range(self.size)) {
                t[i] = self.c[i];
            }
            for (int i in range(other.size())) {
                t[self.size + i] = other.c[i];
            }
            self.c = t;
            self.size += other.size();
            self.real_size = self.size + 4;
        } else {
            for (int i in range(other.size())) {
                self.c[self.size + i] = other.c[i];
            }
            self.size += other.size();
        }
        return;
    }

    method public T pop(self, int n) {
        self.size -= 1;
        return self.c[n];
    }

    method public void remove(self, T value) {
        for (int i in range(self.size)) {
            if (self.c[i] == value) {
                self.size -= 1;
                for (int j in range(i, self.size)) {
                    self.c[j] = self.c[j + 1];
                }
                return;
            }
        }
        raise(Except("Value not found"));
    }

    method public int index(self, T value) {
        for (int i in range(self.size)) {
            if (self.c[i] == value) {
                return i;
            }
        }
        return -1;
    }

    method public int count(self, T value) {
        var int count = 0;
        for (int i in range(self.size)) {
            if (self.c[i] == value) {
                count += 1;
            }
        }
        return count;
    }

    method public void reverse(self) {
        var arr<T> t = arr<T>(self.real_size);
        for (int i in range(self.size)) {
            t[i] = self.c[self.size - 1 - i];
        }
        self.c = t;
        return;
    }

    method public void sort(self) {
        for (int i in range(self.size - 1)) {
            for (int j in range(0, self.size - i - 1)) {
                if (self.c[j] > self.c[j + 1]) {
                    var T t = self.c[j];
                    self.c[j] = self.c[j + 1];
                    self.c[j + 1] = t;
                }
            }
        }
    }
}
