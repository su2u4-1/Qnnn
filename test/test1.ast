program (name: /home/su2u4/code/Qnnn/./test/test1.qn, )
{
    import (alias: stdlib, name: list, ) {}
    class (name: Test, )
    {
        typevar () {}
        declare_attr (kind: attr, modifier: local, name: l, )
        {
            type (name: list, )
            {
                type (name: int, ) {}
            }
        }
        method (kind: private, name: init, )
        {
            type (name: operator, ) {}
            typevar () {}
            args_declare ()
            {
                declare_arg (name: l, tuple: false, )
                {
                    type (name: list, )
                    {
                        type (name: int, ) {}
                    }
                }
            }
            statements ()
            {
                expression ()
                {
                    term (type: variable, )
                    {
                        variable (state: attr, )
                        {
                            variable (state: false, )
                            {
                                name (name: self, ) {}
                            }
                            name (name: l, ) {}
                        }
                    }
                    term (type: variable, )
                    {
                        variable (state: false, )
                        {
                            name (name: l, ) {}
                        }
                    }
                    operator (value: =, ) {}
                }
                return () {}
            }
        }
        method (kind: public, name: get, )
        {
            type (name: int, ) {}
            typevar () {}
            args_declare ()
            {
                declare_arg (name: index, tuple: false, )
                {
                    type (name: int, ) {}
                }
            }
            statements ()
            {
                return ()
                {
                    expression ()
                    {
                        term (type: variable, )
                        {
                            variable (state: index, )
                            {
                                variable (state: attr, )
                                {
                                    variable (state: false, )
                                    {
                                        name (name: self, ) {}
                                    }
                                    name (name: l, ) {}
                                }
                                expression ()
                                {
                                    term (type: variable, )
                                    {
                                        variable (state: false, )
                                        {
                                            name (name: index, ) {}
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        method (kind: private, name: add, )
        {
            type (name: operator, ) {}
            typevar () {}
            args_declare ()
            {
                declare_arg (name: other, tuple: false, )
                {
                    type (name: list, )
                    {
                        type (name: int, ) {}
                    }
                }
            }
            statements ()
            {
                for (else: false, label: for, name: i, )
                {
                    type (name: int, ) {}
                    expression ()
                    {
                        term (type: variable, )
                        {
                            variable (state: false, )
                            {
                                name (name: other, ) {}
                            }
                        }
                    }
                    statements ()
                    {
                        expression ()
                        {
                            term (type: call, )
                            {
                                call ()
                                {
                                    variable (state: attr, )
                                    {
                                        variable (state: attr, )
                                        {
                                            variable (state: false, )
                                            {
                                                name (name: self, ) {}
                                            }
                                            name (name: l, ) {}
                                        }
                                        name (name: append, ) {}
                                    }
                                    use_typevar () {}
                                    args_call ()
                                    {
                                        expression ()
                                        {
                                            term (type: variable, )
                                            {
                                                variable (state: false, )
                                                {
                                                    name (name: i, ) {}
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
                return () {}
            }
        }
    }
}
