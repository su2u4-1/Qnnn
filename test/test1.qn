import list;

class Test {
    attr list<int> l;
    method operator init(self, list<int> l) {
        self.l = l;
        return;
    }

    method public int get(self, int index) {
        return self.l[index];
    }

    method operator add(self, list<int> other) {
        for (int i in other) {
            self.l.append(i);
        }
        return;
    }
}
