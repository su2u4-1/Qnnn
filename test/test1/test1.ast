program ()
{
    declare_var (kind: var, modifier: local, name: a, )
    {
        type (name: int, ) {}
        expression ()
        {
            term (type: int, value: 0, ) {}
        }
    }
    declare_var (kind: const, modifier: local, name: b, )
    {
        type (name: A, )
        {
            type (name: B, ) {}
        }
        expression ()
        {
            term (type: call, )
            {
                variable (state: false, )
                {
                    name (name: A, ) {}
                }
                use_typevar ()
                {
                    type (name: B, ) {}
                }
                args_call ()
                {
                    expression ()
                    {
                        term (type: int, value: 1, ) {}
                    }
                }
            }
        }
    }
    declare_var (kind: var, modifier: global, name: c, )
    {
        type (name: C, )
        {
            type (name: D, ) {}
            type (name: E, ) {}
        }
        expression ()
        {
            term (type: call, )
            {
                variable (state: false, )
                {
                    name (name: C, ) {}
                }
                use_typevar ()
                {
                    type (name: D, ) {}
                    type (name: E, ) {}
                }
                args_call ()
                {
                    expression ()
                    {
                        term (type: int, value: 2, ) {}
                    }
                }
            }
        }
    }
    declare_var (kind: const, modifier: global, name: d, )
    {
        type (name: arr, )
        {
            type (name: int, ) {}
        }
        expression ()
        {
            term (type: arr, )
            {
                arr ()
                {
                    expression ()
                    {
                        term (type: int, value: 3, ) {}
                    }
                    expression ()
                    {
                        term (type: int, value: 3, ) {}
                    }
                    expression ()
                    {
                        term (type: int, value: 3, ) {}
                    }
                }
            }
        }
    }
    class (name: A, )
    {
        typevar ()
        {
            declare_typevar (name: T, ) {}
        }
        declare_attr (kind: attr, modifier: local, name: e, )
        {
            type (name: T, ) {}
            expression () {}
        }
        declare_attr (kind: static, modifier: local, name: f, )
        {
            type (name: T, ) {}
            expression () {}
        }
        declare_attr (kind: attr, modifier: public, name: g, )
        {
            type (name: T, ) {}
            expression () {}
        }
        declare_attr (kind: static, modifier: public, name: h, )
        {
            type (name: T, ) {}
            expression () {}
        }
        method (modifier: public, name: _init, operator: false, self: true, static: false, )
        {
            type (name: void, ) {}
            typevar () {}
            args_declare ()
            {
                declare_arg (name: i, tuple: false, )
                {
                    type (name: int, ) {}
                }
            }
            statements ()
            {
                expression ()
                {
                    term (type: variable, )
                    {
                        variable (state: attr, )
                        {
                            variable (state: false, )
                            {
                                name (name: self, ) {}
                            }
                            name (name: e, ) {}
                        }
                    }
                    term (type: call, )
                    {
                        variable (state: false, )
                        {
                            name (name: T, ) {}
                        }
                        use_typevar () {}
                        args_call ()
                        {
                            expression ()
                            {
                                term (type: variable, )
                                {
                                    variable (state: false, )
                                    {
                                        name (name: i, ) {}
                                    }
                                }
                            }
                        }
                    }
                    operator (value: =, ) {}
                }
                expression ()
                {
                    term (type: variable, )
                    {
                        variable (state: attr, )
                        {
                            variable (state: false, )
                            {
                                name (name: self, ) {}
                            }
                            name (name: f, ) {}
                        }
                    }
                    term (type: call, )
                    {
                        variable (state: false, )
                        {
                            name (name: T, ) {}
                        }
                        use_typevar () {}
                        args_call ()
                        {
                            expression ()
                            {
                                term (type: variable, )
                                {
                                    variable (state: false, )
                                    {
                                        name (name: i, ) {}
                                    }
                                }
                                term (type: int, value: 1, ) {}
                                operator (value: +, ) {}
                            }
                        }
                    }
                    operator (value: =, ) {}
                }
                expression ()
                {
                    term (type: variable, )
                    {
                        variable (state: attr, )
                        {
                            variable (state: false, )
                            {
                                name (name: self, ) {}
                            }
                            name (name: g, ) {}
                        }
                    }
                    term (type: call, )
                    {
                        variable (state: false, )
                        {
                            name (name: T, ) {}
                        }
                        use_typevar () {}
                        args_call ()
                        {
                            expression ()
                            {
                                term (type: variable, )
                                {
                                    variable (state: false, )
                                    {
                                        name (name: i, ) {}
                                    }
                                }
                                term (type: int, value: 2, ) {}
                                operator (value: +, ) {}
                            }
                        }
                    }
                    operator (value: =, ) {}
                }
                expression ()
                {
                    term (type: variable, )
                    {
                        variable (state: attr, )
                        {
                            variable (state: false, )
                            {
                                name (name: self, ) {}
                            }
                            name (name: h, ) {}
                        }
                    }
                    term (type: call, )
                    {
                        variable (state: false, )
                        {
                            name (name: T, ) {}
                        }
                        use_typevar () {}
                        args_call ()
                        {
                            expression ()
                            {
                                term (type: variable, )
                                {
                                    variable (state: false, )
                                    {
                                        name (name: i, ) {}
                                    }
                                }
                                term (type: int, value: 3, ) {}
                                operator (value: +, ) {}
                            }
                        }
                    }
                    operator (value: =, ) {}
                }
                return () {}
            }
        }
    }
}
