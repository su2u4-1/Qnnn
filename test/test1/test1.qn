/*test parse var declare*/

var int a = 0;
const A<B> b = A<B>(1);
var global C<D, E> c = C<D, E>(2);
const global arr<int> d = [3, 3, 3];

class A<T> {
    attr T e;
    static T f;
    attr public T g;
    static public T h;

    method public void _init(self, int i) {
        self.e = T(i);
        self.f = T(i + 1);
        self.g = T(i + 2);
        self.h = T(i + 3);
        return;
    }
}
