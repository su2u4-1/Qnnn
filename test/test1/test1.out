program({},
    declare_var(
        {kind: var, modifier: local, name: a, },
        type({name: int, }),
        expression({}, term({type: int, value: 0, }))
    ),
    declare_var(
        {kind: const, modifier: local, name: b, },
        type({name: A, }, type({name: B, })),
        expression(
            {},
            term(
                {type: call, },
                variable({name: var, state: false, var: A, }),
                use_typevar({}, type({name: B, })),
                args_call({}, expression({}, term({type: int, value: 1, })))
            )
        )
    ),
    declare_var(
        {kind: var, modifier: global, name: c, },
        type({name: C, }, type({name: D, }), type({name: E, })),
        expression(
            {},
            term(
                {type: call, },
                variable({name: var, state: false, var: C, }),
                use_typevar({}, type({name: D, }), type({name: E, })),
                args_call({}, expression({}, term({type: int, value: 2, })))
            )
        )
    ),
    declare_var(
        {kind: const, modifier: global, name: d, },
        type({name: arr, }, type({name: int, })),
        expression(
            {},
            term(
                {type: arr, },
                arr(
                    {},
                    expression({}, term({type: int, value: 3, })),
                    expression({}, term({type: int, value: 3, })),
                    expression({}, term({type: int, value: 3, }))
                )
            )
        )
    ),
    class(
        {name: A, },
        typevar({}, declare_typevar({name: T, })),
        declare_attr({kind: attr, modifier: local, name: e, }, type({name: T, }), expression({})),
        declare_attr({kind: static, modifier: local, name: f, }, type({name: T, }), expression({})),
        declare_attr({kind: attr, modifier: public, name: g, }, type({name: T, }), expression({})),
        declare_attr({kind: static, modifier: public, name: h, }, type({name: T, }), expression({})),
        method(
            {modifier: public, name: init, operator: false, self: true, static: false, },
            type({name: A, }),
            typevar({}),
            args({}, declare_arg({name: i, tuple: false, }, type({name: int, }))),
            statements(
                {},
                expression(
                    {},
                    term({type: variable, }, variable({name: e, state: attr, }, variable({name: var, state: false, var: self, }))),
                    term(
                        {type: call, },
                        variable({name: var, state: false, var: T, }),
                        use_typevar({}),
                        args_call({}, expression({}, term({type: variable, }, variable({name: var, state: false, var: i, }))))
                    ),
                    operator({value: =, })
                ),
                expression(
                    {},
                    term({type: variable, }, variable({name: f, state: attr, }, variable({name: var, state: false, var: self, }))),
                    term(
                        {type: call, },
                        variable({name: var, state: false, var: T, }),
                        use_typevar({}),
                        args_call(
                            {},
                            expression(
                                {},
                                term({type: variable, }, variable({name: var, state: false, var: i, })),
                                term({type: int, value: 1, }),
                                operator({value: +, })
                            )
                        )
                    ),
                    operator({value: =, })),
                expression(
                    {},
                    term({type: variable, }, variable({name: g, state: attr, }, variable({name: var, state: false, var: self, }))),
                    term(
                        {type: call, },
                        variable({name: var, state: false, var: T, }),
                        use_typevar({}),
                        args_call(
                            {},
                            expression(
                                {},
                                term({type: variable, }, variable({name: var, state: false, var: i, })),
                                term({type: int, value: 2, }),
                                operator({value: +, })
                            )
                        )
                    ),
                    operator({value: =, })),
                expression(
                    {},
                    term({type: variable, }, variable({name: h, state: attr, }, variable({name: var, state: false, var: self, }))),
                    term(
                        {type: call, },
                        variable({name: var, state: false, var: T, }),
                        use_typevar({}),
                        args_call(
                            {},
                            expression(
                                {},
                                term({type: variable, }, variable({name: var, state: false, var: i, })),
                                term({type: int, value: 3, }),
                                operator({value: +, })
                            )
                        )
                    ),
                    operator({value: =, })),
                return({}, expression({}, term({type: variable, }, variable({name: var, state: false, var: self, }))))
            )
        )
    )
)